session/bootstrap.go

```diff
@@ -234,16 +234,16 @@ const (
 
 	// CreateBindInfoTable stores the sql bind info which is used to update globalBindCache.
 	CreateBindInfoTable = `CREATE TABLE IF NOT EXISTS mysql.bind_info (
-		original_sql	TEXT NOT NULL  ,
-      	bind_sql 		TEXT NOT NULL ,
-      	default_db 		TEXT NOT NULL,
-		status 			TEXT NOT NULL,
-		create_time 	TIMESTAMP(3) NOT NULL,
-		update_time 	TIMESTAMP(3) NOT NULL,
-		charset 		TEXT NOT NULL,
-		collation 		TEXT NOT NULL,
-		source 			VARCHAR(10) NOT NULL DEFAULT 'unknown',
-		INDEX sql_index(original_sql(1024),default_db(1024)) COMMENT "accelerate the speed when add global binding query",
+		original_sql VARCHAR(512) NOT NULL,
+		bind_sql VARCHAR(512) NOT NULL,
+		default_db VARCHAR(32) NOT NULL,
+		status VARCHAR(32) NOT NULL,
+		create_time TIMESTAMP(3) NOT NULL,
+		update_time TIMESTAMP(3) NOT NULL,
+		charset VARCHAR(32) NOT NULL,
+		collation VARCHAR(32) NOT NULL,
+		source VARCHAR(10) NOT NULL DEFAULT 'unknown',
+		UNIQUE INDEX sql_db_src_idx(original_sql,default_db,source) COMMENT "accelerate the speed when add global binding query",
```

Comment:
This would break baseline evolution because we now impose the restriction that there is at most one record for evolution binding. Considered that baseline evolution is an incomplete feature, and I can think of no other approaches to fix this concurrency problem, this break is acceptable?

Reply:
shorten the length limitation of `original_sql` and `bind_sql` also breaks the backward compacity. How about defining them in `text` as original, and define a prefix index on these fields?

Reply:
> for TEXT / BLOB type columns, if we want to build index on them, the column must be specified with a prefix length, which prevents planner from choosing PointGet / BatchPointGet for the `select ... for update` query. While according to https://docs.pingcap.com/tidb/dev/pessimistic-transaction#behaviors, only PointGet / BatchPointGet can lock the keys even if there is no record satisfying the filter. For normal IndexScan, only the records satisfying the filter would be locked, thus it is possible to have duplicate records in mysql.bind_info if 2 `CREATE BINDING` queries are executed simultaneously on 2 tidb instances respectively.